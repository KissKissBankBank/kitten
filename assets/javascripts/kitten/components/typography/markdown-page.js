import React, { Fragment } from 'react'
import styled from 'styled-components'
import Markdown from 'react-markdown'
import PropTypes from 'prop-types'
import slugify from 'slugify'
import { Marger } from '../../components/layout/marger'
import { Title } from '../../components/typography/title'
import { Paragraph } from '../../components/typography/paragraph'
import { BulletList } from '../../components/lists/bullet-list'
import { LinkList } from '../../components/links/link-list'
import { HorizontalStroke } from '../../components/layout/horizontal-stroke'
import { scrollTo } from '../../helpers/utils/scroll-to'
import { pxToRem } from '../../helpers/utils/typography'
import { ScreenConfig } from '../../constants/screen-config'

export const MarkdownPage = props => (
  <Markdown
    source={props.content}
    renderers={markdownRenderers}
    escapeHtml={false}
  />
)

MarkdownPage.propTypes = {
  content: PropTypes.string.isRequired,
}

const MarkdownParagraph = props => {
  return (
    <Marger bottom="2.6">
      <Paragraph modifier="primary" margin={false}>
        {props.children}
      </Paragraph>
    </Marger>
  )
}

MarkdownParagraph.propTypes = {
  children: PropTypes.node.isRequired,
}

const MarkdownBlockquote = props => {
  const items = props.children.map(item => ({
    key: item.key,
    item: item.props.children[0],
  }))

  return (
    <Marger top="4" bottom="4">
      <LinkListstyle color="dark" items={items} className="kiss-LinkList" />
    </Marger>
  )
}

MarkdownBlockquote.propTypes = {
  children: PropTypes.node.isRequired,
}

const MarkdownListComponent = props => {
  const items = props.children.map(item => ({
    key: item.key,
    item: item.props.children[0],
  }))

  return (
    <Marger top="3" bottom="3">
      <BulletListStyle className="kiss-BulletList" big items={items} />
    </Marger>
  )
}

MarkdownListComponent.propTypes = {
  children: PropTypes.node.isRequired,
}

const handleLinkClick = url => {
  if (url.match(/^#/)) {
    scrollTo(document.getElementById(url.replace('#', '')), 500)
  }
}

const MarkdownLink = props => {
  return (
    <a
      href={props.href}
      className="k-Paragraph__link"
      onClick={() => {
        handleLinkClick(props.href)
      }}
    >
      {props.children}
    </a>
  )
}

MarkdownLink.propTypes = {
  href: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

const MarkdownHeading = ({ ...props }) => {
  let modifier
  let tag

  switch (props.level) {
    case 1:
      modifier = 'secondary'
      tag = 'h1'
      break
    case 2:
      modifier = 'quaternary'
      tag = 'h2'
      break
    case 3:
      modifier = 'quinary'
      tag = 'h3'
      break
    default:
      modifier = 'senary'
      tag = 'h4'
      break
  }

  const renderHorizontalStroke = () => {
    if (props.level > 1) return

    return (
      <Marger top="2" bottom="4">
        <HorizontalStroke size="big" />
      </Marger>
    )
  }

  const id = slugify(props.children[0].props.value, {
    lower: true,
    remove: /[']/g,
  })

  return (
    <Fragment>
      <TitleMarger>
        <Title modifier={modifier} tag={tag} id={id}>
          {props.children}
        </Title>
      </TitleMarger>
      {renderHorizontalStroke()}
    </Fragment>
  )
}

MarkdownHeading.defaultProps = {
  level: null,
}

MarkdownHeading.propTypes = {
  level: PropTypes.number,
  children: PropTypes.node.isRequired,
}

const markdownRenderers = {
  paragraph: MarkdownParagraph,
  link: MarkdownLink,
  heading: MarkdownHeading,
  blockquote: MarkdownBlockquote,
}

const LinkListstyle = styled(LinkList)`
  margin: 0;

  li {
    text-align: left;
  }
`

const BulletListStyle = styled(BulletList)`
  li: {
    lineheight: '1.8rem';
  }
`

const TitleMarger = styled(Marger)`
  margin-top: ${pxToRem(100)};
  margin-bottom: ${pxToRem(20)};

  @media (max-width: ${ScreenConfig.S.max}px) {
    margin-top: ${pxToRem(80)};
  }
`
