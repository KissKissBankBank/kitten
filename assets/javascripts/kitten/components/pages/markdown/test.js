import React from 'react'
import renderer from 'react-test-renderer'
import { MarkdownPage } from './markdown-page'

describe('<MarkdownPage />', () => {
  let component

  beforeEach(() => {
    component = renderer.create(<MarkdownPage content="foobar" />).toJSON()
  })

  it('matches with snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})
