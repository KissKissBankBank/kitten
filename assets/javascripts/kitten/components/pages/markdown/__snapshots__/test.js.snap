// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MarkdownPage /> with \`paragraphModifier\` props matches with snapshot 1`] = `
<div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="h1"
    >
      H1
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h2
      className="title__StyledTitle-sc-46lshq-0 jSlMJn"
      id="h2"
    >
      H2
    </h2>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h3
      className="title__StyledTitle-sc-46lshq-0 kWyuxU"
      id="h3"
    >
      H3
    </h3>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="emphasis"
    >
      Emphasis
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 loOPNX"
  >
    <p
      className="k-Paragraph k-Paragraph--primary k-Paragraph--withoutMargin"
    >
      Emphasis, aka italics, with 
      <em>
        asterisks
      </em>
       or 
      <em>
        underscores
      </em>
      .
Strong emphasis, aka bold, with 
      <strong>
        asterisks
      </strong>
       or 
      <strong>
        underscores
      </strong>
      .
Combined emphasis with 
      <strong>
        asterisks and 
        <em>
          underscores
        </em>
      </strong>
      .
Strikethrough uses two tildes. ~~Scratch this.~~
    </p>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="list"
    >
      List
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kWeiWM"
      >
        Unordered list can use asterisks
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kWeiWM"
      >
        Or minuses
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kWeiWM"
      >
        Or pluses
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="link"
    >
      Link
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 loOPNX"
  >
    <p
      className="k-Paragraph k-Paragraph--primary k-Paragraph--withoutMargin"
    >
      <a
        className="k-Paragraph__link"
        href="https://www.google.com"
      >
        I'm an inline-style link
      </a>
    </p>
  </div>
</div>
`;

exports[`<MarkdownPage /> with children matches with snapshot 1`] = `
<div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="h1"
    >
      H1
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h2
      className="title__StyledTitle-sc-46lshq-0 jSlMJn"
      id="h2"
    >
      H2
    </h2>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h3
      className="title__StyledTitle-sc-46lshq-0 kWyuxU"
      id="h3"
    >
      H3
    </h3>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="emphasis"
    >
      Emphasis
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 loOPNX"
  >
    <p
      className="k-Paragraph k-Paragraph--quaternary k-Paragraph--withoutMargin"
    >
      Emphasis, aka italics, with 
      <em>
        asterisks
      </em>
       or 
      <em>
        underscores
      </em>
      .
Strong emphasis, aka bold, with 
      <strong>
        asterisks
      </strong>
       or 
      <strong>
        underscores
      </strong>
      .
Combined emphasis with 
      <strong>
        asterisks and 
        <em>
          underscores
        </em>
      </strong>
      .
Strikethrough uses two tildes. ~~Scratch this.~~
    </p>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="list"
    >
      List
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kXnRej"
      >
        Unordered list can use asterisks
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kXnRej"
      >
        Or minuses
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 gKSWLw"
  >
    <ul
      className="bullet-list__StyledBulletList-rdyx84-0 eBDmYd list__BulletListStyle-sc-1hya2vn-0 fMlCJG"
    >
      <li
        className="bullet-list__StyledItem-rdyx84-1 kXnRej"
      >
        Or pluses
      </li>
    </ul>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 kboIqb"
  >
    <h1
      className="title__StyledTitle-sc-46lshq-0 kXPoIm"
      id="link"
    >
      Link
    </h1>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 PUhWi"
  >
    <span
      className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 cQtxoj"
    />
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 loOPNX"
  >
    <p
      className="k-Paragraph k-Paragraph--quaternary k-Paragraph--withoutMargin"
    >
      <a
        className="k-Paragraph__link"
        href="https://www.google.com"
      >
        I'm an inline-style link
      </a>
    </p>
  </div>
</div>
`;
